Fixed issues with openmrs setup process

jecihjoy opened the pull request:
<!-- This is based on openmrs-cord PR template. -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
<!--- If your PR is related to an issue , please mention it here. -->
<!--- It is generally a good practice to first file an issue with enough
  context and reference it in the PR, but if you don't have that, please remove
  this section. -->
Fixed current issues that occurs when setting up OpenMrs.
Generated test data with 1000 patients
Packaged both fhir2 and atom feed module with the docker image

## E2E test
<!-- There are different scenarios for using the tools in this repo; please 
  help your reviewers by describing how you have e2e tested your change. -->
TESTED:

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My IDE is configured to follow the
  [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of
  this project.

  No? Unsure? ->
  [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE),
  format the code and add the changes with `git add . && git commit --amend`

- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review
  [Java](https://google.github.io/styleguide/javaguide.html) and
  [Python](https://google.github.io/styleguide/pyguide.html) style guides. Note,
  when in conflict, OpenMRS style guide overrules.

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your
  responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

kimaina commented on .travis.yml:
```diff
@@ -2,7 +2,9 @@ language: java
   - openjdk11
-
-script:
-  mvn clean install
```
did you delete `mvn clean install` by mistake?

kimaina replied:
thanks for adding back

kimaina reviewed (commented):
Thanks for working on this. How many patients/obs/encounters did you generate?

jecihjoy commented:
> Thanks for working on this. How many patients/obs/encounters did you generate?

Patients --> 1000
Encounters --> 11478
Obs --> 94736

This data will be updated from the work going on [here](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/issues/59) to reflect what's expected

bashir2 commented on .travis.yml:
```diff
@@ -2,7 +2,11 @@ language: java
+services:
+  - docker
+before_install:
+  - travis_wait docker-compose -f openmrs-compose.yaml up -d
```
Should we do a `curl` too? IIUC the first request to the RefApp UI will trigger some initialization that can take some time as well, right? And if that is the case then I wonder if we need a `travis_wait` for that too. Another preferred solution is to put the `docker-compose` command, the `curl` request after that, and any other initialization step, in a script `foo.sh` and just `travis_wait foo.sh`. That way people can repeat those steps in their local env. easily. WDYT?

jecihjoy replied:
Yes I think moving these steps that takes long to run to a separate file makes sense so that we can only run ```travis_wait``` once

bashir2 replied:
Following up on this thread: Should we also move the `curl` command to the new `travis.sh`?

jecihjoy replied:
@bashir2 adding this check to `travis.sh` means that both `docker-compose -f openmrs-compose.yaml up -d` and the curl check will be executed in the same lifecycle `before_install`. Having both in the same life cycle the curl command is executed before the container is up hence it fails. It's not possible to use travis lifecycle inside the sh file...

bashir2 replied:
I see, okay, never mind then.

bashir2 commented on openmrs-compose.yaml:
```diff
@@ -37,7 +36,7 @@ services:
   openmrs:
     restart: "always"
-    image: openmrs/openmrs-reference-application-distro:nightly
+    image: ampathke/openmrs-analytics-distro:latest
```
How easy is it for people outside AMPATH to update this image? If it is not easy, then I would prefer an `openmrs` image. Also maybe instead of `nightly` or `latest` we should use a specific tag, e.g., `2.10` such that if `latest` moves, it does not break our e2e tests.

jecihjoy replied:
@bashir2 this is an openmrs image only that it has been hosted in the ampath org in docker hub
It's similar to ```openmrs-reference-application-distro:qa``` plus ```atom feed module```

bashir2 replied:
I see; yeah I think moving it under openmrs is better. BTW, where is the `Dockerfile` for this image? Should we add that to this code repository as well?

bashir2 reviewed (commented):
Thanks @jecihjoy for working on the e2e framework; this is a major dev. pain point.

jecihjoy commented:
@bashir2 I’ve had a lot of back and forth discussions with @dkayiwa and @cintiadr to have them publish for us a stable openmrs image packaged with latest versions of atom feed and fhir2 modules but we haven’t finalized on this

jecihjoy commented:
The ideal docker image to use would be the one managed by the community so that we can get the benefits of frequent updates of the openmrs platform and the modules. However non of these images has been packaged by both fhir2 and atom feed modules

jecihjoy commented:
That said I think we can start by using the image hosted in `Ampath` org and once we agree with @dkayiwa and @cintiadr on having a stable openmrs docker image tagged analytics  hosted in the openmrs org then we’ll update the project’s image. [Here](https://github.com/jecihjoy/openmrs-fhir-analytics) is the repo containing the Dockerfile

bashir2 commented:
Thanks for looking into this @jecihjoy CIL:

> That said I think we can start by using the image hosted in `Ampath` org and once we agree with @dkayiwa and @cintiadr on having a stable openmrs docker image tagged analytics hosted in the openmrs org then we’ll update the project’s image. [Here](https://github.com/jecihjoy/openmrs-fhir-analytics) is the repo containing the Dockerfile and the steps to generate it are as follows
>

Having the docker image under `ampath` temporarily is fine. Can you please file a bug to track the fix you are suggesting? We should probably document the steps somewhere here in this repo.

BTW, I could not find the Dockerfile, can you please directly link that file (your link is to the `master` branch) or the commit containing it?
 
> 1. Setup latest OpenMrs distro locally `mvn openmrs-sdk:setup -DserverId=openmrs-analytics`
> 2. Add latest versions of `fhir2 and atom feed modules` and ensure everything is running as expected
> 3. Build a docker image using openmrs SDK, this step requires you to provide a path to the openmrs-distro.properties file of your openmrs instance `mvn openmrs-sdk:build-distro -Ddistro=/{path}/openmrs-distro.properties`
> 4. Build the docker image
>    `cd openmrs-analytics/web`
>    `docker build -t openmrs/openmrs-reference-application-distro:analytics .`

If we stop supporting the Atom Feed based pipeline, do we still need this custom image or will the standard platform release (containing `fhir2`) be enough for us?

jecihjoy commented:
> Having the docker image under `ampath` temporarily is fine. Can you please file a bug to track the fix you are suggesting?

@bashir2  this PR addresses the issues reported [here](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/issues/38) because of the unstability of the initial docker image that we were using.



> We should probably document the steps somewhere here in this repo.

I would be happy to move these steps to an agreed position. 

> BTW, I could not find the Dockerfile, can you please directly link that file (your link is to the `master` branch) or the commit containing it?

The docker image is in the [web folder](https://github.com/jecihjoy/openmrs-docker-sdk/blob/master/web/Dockerfile) in the master branch. The docker image must be packaged by both openmrs platform and modules hence the many folders.

> If we stop supporting the Atom Feed based pipeline, do we still need this custom image or will the standard platform release (containing `fhir2`) be enough for us?

The answer is yes. The openmrs distro images that has fhir2 include `openmrs-reference-application-distro:qa` and `openmrs-reference-application-distro:nightly` however I cannot acertain the stability of these images. However my suggestion is we use `ampathke/openmrs-reference-application-distro:analytics` until `openmrs-reference-application-distro:latest` image is updated with the latest modules and fhir2 module added

jecihjoy commented:
> That said I think we can start by using the image hosted in `Ampath` org and once we agree with @dkayiwa and @cintiadr on having a stable openmrs docker image tagged analytics hosted in the openmrs org then we’ll update the project’s image. [Here](https://github.com/jecihjoy/openmrs-fhir-analytics) is the repo containing the Dockerfile and the steps to generate it are as follows
> 
> 1. Setup latest OpenMrs distro locally `mvn openmrs-sdk:setup -DserverId=openmrs-analytics`
> 2. Add latest versions of `fhir2 and atom feed modules` and ensure everything is running as expected
> 3. Build a docker image using openmrs SDK, this step requires you to provide a path to the openmrs-distro.properties file of your openmrs instance `mvn openmrs-sdk:build-distro -Ddistro=/{path}/openmrs-distro.properties`
> 4. Build the docker image
>    `cd openmrs-analytics/web`
>    `docker build -t openmrs/openmrs-reference-application-distro:analytics .`

How about adding a link in readme that points to this comment and we'll remove this once we start using the standard image

pmanko commented:
I'm not sure if this is relevant or not to this PR, but when testing the `streaming-atomfeed` docker setup using this branch's openmrs setup, I had to clear the `event_records` table after updating the atomfeed configs, since the old records did not include the correct resource path for the `fhir2` module. Only new records created after updating the `atomfeed` config had the correct resource path. 

Were these records generated on an instance running without the updated config file? 

And on a related note, would we be able to bundle an atomfeed module version with the updated config file to further simplify the openmrs setup process?

bashir2 commented:
> I'm not sure if this is relevant or not to this PR, but when testing the `streaming-atomfeed` docker setup using this branch's openmrs setup, I had to clear the `event_records` table after updating the atomfeed configs, since the old records did not include the correct resource path for the `fhir2` module. Only new records created after updating the `atomfeed` config had the correct resource path.

@pmanko when you say "this branch" do you mean with this PR included? Because there is the known issue #38 which this PR is supposed to fix.

If you are talking about the FHIR resources being wrong with this PR included, then @jecihjoy can you please take a look at this? I mean making sure that the event records that are created by the Atom Feed module use the new `fhir2` module URLs. I think you can check this by looking at the `object` column of the `event_records` table (please correct me @pmanko, if that is not the case).

> 
> Were these records generated on an instance running without the updated config file?
> 
> And on a related note, would we be able to bundle an atomfeed module version with the updated config file to further simplify the openmrs setup process?

Yeah, that should be the case if we want to include the Atom Feed module in the docker image; @jecihjoy can you please check/confirm?

bashir2 commented:
> > That said I think we can start by using the image hosted in `Ampath` org and once we agree with @dkayiwa and @cintiadr on having a stable openmrs docker image tagged analytics hosted in the openmrs org then we’ll update the project’s image. [Here](https://github.com/jecihjoy/openmrs-fhir-analytics) is the repo containing the Dockerfile and the steps to generate it are as follows
> > 
> > 1. Setup latest OpenMrs distro locally `mvn openmrs-sdk:setup -DserverId=openmrs-analytics`
> > 2. Add latest versions of `fhir2 and atom feed modules` and ensure everything is running as expected
> > 3. Build a docker image using openmrs SDK, this step requires you to provide a path to the openmrs-distro.properties file of your openmrs instance `mvn openmrs-sdk:build-distro -Ddistro=/{path}/openmrs-distro.properties`
> > 4. Build the docker image
> >    `cd openmrs-analytics/web`
> >    `docker build -t openmrs/openmrs-reference-application-distro:analytics .`
> 
> How about adding a link in readme that points to this comment and we'll remove this once we start using the standard image

Sounds good.

bashir2 commented:
> > Having the docker image under `ampath` temporarily is fine. Can you please file a bug to track the fix you are suggesting?
> 
> @bashir2 this PR addresses the issues reported [here](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/issues/38) because of the unstability of the initial docker image that we were using.

You are right, let's track it there; I just reopened the issue.

> 
> > We should probably document the steps somewhere here in this repo.
> 
> I would be happy to move these steps to an agreed position.

Let's create a new `doc/` folder and add a `developers.md` or something like that. We should trim some content for the current `README.md` and move it there too (this is mostly a note to myself).

> 
> > BTW, I could not find the Dockerfile, can you please directly link that file (your link is to the `master` branch) or the commit containing it?
> 
> The docker image is in the [web folder](https://github.com/jecihjoy/openmrs-docker-sdk/blob/master/web/Dockerfile) in the master branch. The docker image must be packaged by both openmrs platform and modules hence the many folders.

Okay, thanks; just documenting the process for now is fine. I am not sure if we need to support Atom Feed module long term and in that case it is best to use an OpenMRS official image release.

> 
> > If we stop supporting the Atom Feed based pipeline, do we still need this custom image or will the standard platform release (containing `fhir2`) be enough for us?
> 
> The answer is yes. The openmrs distro images that has fhir2 include `openmrs-reference-application-distro:qa` and `openmrs-reference-application-distro:nightly` however I cannot acertain the stability of these images. However my suggestion is we use `ampathke/openmrs-reference-application-distro:analytics` until `openmrs-reference-application-distro:latest` image is updated with the latest modules and fhir2 module added

Okay, sounds good.

bashir2 commented:
BTW, @jecihjoy you might have missed [this comment](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/pull/66#discussion_r534688655) because of GitHub sub-optimal comment UI. I think we can merge this once `travis.sh` is updated.

pmanko commented:
> @pmanko when you say "this branch" do you mean with this PR included? Because there is the known issue #38 which this PR is supposed to fix.

Yes, I checked out the branch for the PR and started up the `openmrs-compose.yml` containers before switching back to the branch I was working on. 

> If you are talking about the FHIR resources being wrong with this PR included, then @jecihjoy can you please take a look at this? I mean making sure that the event records that are created by the Atom Feed module use the new `fhir2` module URLs. I think you can check this by looking at the `object` column of the `event_records` table (please correct me @pmanko, if that is not the case).

I looked in https://github.com/jecihjoy/openmrs-fhir-analytics/blob/b38/utils/dbdump/backup.sql, and it seems the `event_record` entries in the file include the old paths in the `object` column.

jecihjoy commented:
>  it seems the `event_record` entries in the file include the old paths in the `object` column.
@pmanko What do you mean by this?

bashir2 commented:
> > it seems the `event_record` entries in the file include the old paths in the `object` column.
> 
> @pmanko What do you mean by this?

@jecihjoy when you set up both MySQL and OpenMRS docker images, what do you see in `event_record ` table:
```
SELECT * FROM event_records LIMIT 10;
```
In the `object` column what are the FHIR resource URLs? Are they `/ws/fhir2/R4/...` or do they use the old `fhir` module (which I think starts with `/ws/fhir/`)? If it is the latter then it means the Atom Feed module was misconfigured at the time that the event were generated. In particular, [this step for its config update](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics#add-fhir2-module-to-openmrs-and-update-atom-feed-config) was not done. Does that make sense?

@pmanko please clarify further if I misunderstood what you mean.

jecihjoy commented:
> In particular, [this step for its config update](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics#add-fhir2-module-to-openmrs-and-update-atom-feed-config) was not done. Does that make sense?

Thank you @pmanko and @bashir2 for identifying this. I will configure atom feed installation and update the PR

jecihjoy commented:
@bashir2 @pmanko I have updated the PR with the feedbacks

> Were these records generated on an instance running without the updated config file?

I realized the atom feed configuration is not persisted between openmrs server restarts, and because the demo data is being generated at run time the data will be generated with the old atom feed configs

> And on a related note, would we be able to bundle an atomfeed module version with the updated config file to further simplify the openmrs setup process?

To prevent the above from happening and to respond to this request, I replaced the atom feed config file with the new configuration. Here is the [commit](https://github.com/jecihjoy/openmrs-module-atomfeed/commit/8db08aa4d11073bb9c942005ff06336b2de62061) for this. WDYT about this?

pmanko commented:
Thanks @jecihjoy I'll test it out on my side, but looks like it should work. 

On a related note, I might have some fixes to the atomfeed module as well, so it might be useful to discuss perhaps a new release of the module. 

> To prevent the above from happening and to respond to this request, I replaced the atom feed config file with the new configuration. Here is the [commit](https://github.com/jecihjoy/openmrs-module-atomfeed/commit/8db08aa4d11073bb9c942005ff06336b2de62061) for this. WDYT about this?

The only issue I see is the need to update this config to use the `R4` endpoints instead, since this PR has been merged: https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/pull/68

bashir2 reviewed (approved):
LGTM; thanks @jecihjoy for making the changes. @kimaina can merge this while I am away.

kimaina commented:
> LGTM; thanks @jecihjoy for making the changes. @kimaina can merge this while I am away.

Sure @bashir2 ! @jecihjoy let me know once you revert!

jecihjoy commented:
> The only issue I see is the need to update this config to use the `R4` endpoints instead, since this PR has been merged: #68

@kimaina I have updated the PR with this request and it should be ready for merging now

kimaina reviewed (approved):
Thanks for working on this!

kimaina reviewed (approved):
sorry I forgot to merge this!
